File: ui_tests/test_ui.py
import asyncio
from playwright.async_api import async_playwright

async def run(playwright):
    browser = await playwright.chromium.launch(headless=True)
    context = await browser.new_context(viewport={"width": 1280, "height": 800})
    page = await context.new_page()

    page.on("console", lambda msg: print(f"[Browser] {msg.text}"))
    page.on("pageerror", lambda err: print(f"[Browser Error] {err}"))

    try:
        print("1. Navigating to Dashboard...")
        await page.goto("http://localhost:8080")
        await page.wait_for_selector("#nav-dashboard")
        
        print("2. Checking Dashboard elements...")
        title = await page.locator("#page-title").inner_text()
        
        print("3. Testing Navigation to New Job Wizard...")
        await page.click("#nav-new-job")
        
        # Wait for the tab content to become visible instead of title matching exactly
        await page.wait_for_selector("#tab-new-job:not(.hidden)")
        
        print("4. Filling out New Job Form...")
        await page.fill("#input-target", "toko kopi")
        
        await page.wait_for_timeout(500)
        await page.fill("#select-province", "DKI Jakarta")
        await page.wait_for_selector("#dropdown-province .combobox-item")
        await page.click("#dropdown-province .combobox-item:first-child")
        
        await page.wait_for_timeout(500)
        await page.fill("#select-city", "Kota Jakarta Pusat")
        await page.wait_for_selector("#dropdown-city .combobox-item")
        await page.click("#dropdown-city .combobox-item:first-child")
        
        print("5. Generating queries...")
        await page.click("button[onclick='generateQueries()']")
        await page.wait_for_timeout(500)
        queries = await page.locator("#input-queries").input_value()
        if not queries:
            raise Exception("Queries were not generated!")
        print("Generated queries:", queries.replace('\n', ', '))
        
        print("6. Submitting the job...")
        await page.select_option("#select-headless", "true")
        await page.click("#btn-submit-job")
        
        print("7. Waiting for redirect to Dashboard and job to start...")
        await page.wait_for_selector("#tab-dashboard:not(.hidden)", timeout=10000)
        
        print("8. Verifying job status updates...")
        await page.wait_for_selector("#job-status-badge")
        badge = await page.locator("#job-status-badge").inner_text()
        print(f"Job status badge: {badge}")
        
        print("9. Testing Logs Tab...")
        await page.click("#nav-logs")
        await page.wait_for_selector("#tab-logs:not(.hidden)")
        await page.wait_for_timeout(1000)
        
        print("10. Testing Records Tab...")
        await page.click("#nav-records")
        await page.wait_for_selector("#tab-records:not(.hidden)")
        
        print("11. Testing Map Tab...")
        await page.click("#nav-map")
        await page.wait_for_selector("#tab-map:not(.hidden)")
        
        print("\nAll UI tests passed successfully!")
    except Exception as e:
        print(f"\nUI Test Failed: {e}")
    finally:
        await browser.close()

async def main():
    async with async_playwright() as playwright:
        await run(playwright)

if __name__ == "__main__":
    asyncio.run(main())

File: ui_tests/test_ui.js
const puppeteer = require('puppeteer');

(async () => {
    console.log("Starting UI tests...");
    const browser = await puppeteer.launch({ headless: true });
    const page = await browser.newPage();
    
    // Set viewport
    await page.setViewport({ width: 1280, height: 800 });
    
    let logs = [];
    page.on('console', msg => logs.push(`[Browser] ${msg.text()}`));
    page.on('pageerror', err => logs.push(`[Browser Error] ${err.toString()}`));
    
    try {
        console.log("1. Navigating to Dashboard...");
        await page.goto('http://localhost:8080');
        await page.waitForSelector('#nav-dashboard');
        
        console.log("2. Checking Dashboard elements...");
        const title = await page.$eval('#page-title', el => el.innerText);
        if(title !== 'Scraping Dashboard') throw new Error("Dashboard title mismatch: " + title);
        
        console.log("3. Testing Navigation to New Job Wizard...");
        await page.click('#nav-new-job');
        await page.waitForFunction(() => document.getElementById('page-title').innerText === 'New Job Setup Wizard');
        
        console.log("4. Filling out New Job Form...");
        await page.type('#input-target', 'toko kopi');
        
        // Wait for province to load
        await page.waitForTimeout(500);
        await page.type('#select-province', 'DKI Jakarta');
        await page.waitForSelector('#dropdown-province .combobox-item');
        await page.click('#dropdown-province .combobox-item');
        
        // Wait for city to load
        await page.waitForTimeout(500);
        await page.type('#select-city', 'Kota Jakarta Pusat');
        await page.waitForSelector('#dropdown-city .combobox-item');
        await page.click('#dropdown-city .combobox-item');
        
        console.log("5. Generating queries...");
        await page.click('button[onclick="generateQueries()"]');
        await page.waitForTimeout(500);
        const queries = await page.$eval('#input-queries', el => el.value);
        if(!queries) throw new Error("Queries were not generated!");
        console.log("Generated queries:", queries);
        
        console.log("6. Submitting the job...");
        // Set headless to ON
        await page.select('#select-headless', 'true');
        await page.click('#btn-submit-job');
        
        console.log("7. Waiting for redirect to Dashboard and job to start...");
        await page.waitForFunction(() => document.getElementById('page-title').innerText === 'Scraping Dashboard', {timeout: 10000});
        
        // Wait for progress to show
        console.log("8. Verifying job status updates...");
        await page.waitForSelector('#job-status-badge');
        const badge = await page.$eval('#job-status-badge', el => el.innerText);
        console.log("Job status badge:", badge);
        
        console.log("9. Testing Logs Tab...");
        await page.click('#nav-logs');
        await page.waitForFunction(() => document.getElementById('page-title').innerText === 'Logs Viewer');
        await page.waitForTimeout(2000); // let logs stream
        
        console.log("10. Testing Records Tab...");
        await page.click('#nav-records');
        await page.waitForFunction(() => document.getElementById('page-title').innerText === 'Scraper Records Explorer');
        
        console.log("11. Testing Map Tab...");
        await page.click('#nav-map');
        await page.waitForFunction(() => document.getElementById('page-title').innerText === 'Scraped Locations Map');
        
        console.log("All UI tests passed!");
    } catch(err) {
        console.error("UI Test Failed:", err);
        console.log("Browser Logs:", logs);
    } finally {
        await browser.close();
    }
})();
